### DISCUSSION TITLE
Direct access to pixel memory on ROI image

### DISCUSSION THREAD

**Discussion by sn4k3** [emgucv/emgucv#593] [Q&A]
Currently i have this:

```C#
        public static unsafe Span<byte> GetDataByteSpan(this Mat mat)
        {
            return new Span<byte>(mat.DataPointer.ToPointer(), mat.GetLength());
        }
```

It works fine for pixel manipulation, but when executing this on a ROI mat, set pixels below first row will continue to sequential set pixels on original Mat. That is understandable, the **DataPointer** points to first pixel on the ROI and then any pixel > than first row will be set outside ROI, here an example, where blue square is ROI and noise pixels are set via this span method:

![image](https://user-images.githubusercontent.com/113281/135780188-42162c67-770b-4d17-812f-c57a281099a5.png)

It works as should be, since ROI mat is a submatrix and not a new matrix with new data and pointers. So i understand that.
My question: Is there any easy way to use memory access to read/set pixels correctly on ROI mat using **sequential** indexing? (Without jumps)

**Comment by emgucv** [comment]
The data in the ROI are not fully sequential, each row is separated by a "rowSpan" (original image row size in bytes).

It looks like Span2D<T> may help:
https://docs.microsoft.com/en-us/windows/communitytoolkit/high-performance/span2d
but have I have tried.

Based on the documentaion I would try to use this constructor:
https://docs.microsoft.com/en-us/dotnet/api/microsoft.toolkit.highperformance.span2d-1.-ctor?view=win-comm-toolkit-dotnet-7.0#Microsoft_Toolkit_HighPerformance_Span2D_1__ctor_System_Void__System_Int32_System_Int32_System_Int32_

You will need to provide the correct "pitch" parameter, for gray scale single channel image this should be the same as the rowSpan.

**Reply by sn4k3** [reply]
Nice! I was looking for something like that into C# Span but never found it... I will try that for sure!
Btw, given a Mat that is a submatrix, is it possible to get it parent original matrix?
Thanks
